### ISSUE TITLE
trial --dry-run fails with real test_suites, because pyunit suites don't have names

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2024]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| spiv reported|
|-|-|
|Trac ID|trac#2024|
|Type|defect|
|Created|2006-08-29 06:42:42Z|

pyunit doesn't define a 'name' attribute on TestSuites.  If you e.g. remove the `ts.name = 'CustomSuite'` line from the mockcustomtestsuite.py and try to --dry-run it, things break.

<details><summary>Searchable metadata</summary>

```
trac-id__2024 2024
type__defect defect
reporter__spiv spiv
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1156833762000000 1156833762000000
changetime__1166770119000000 1166770119000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_1"></a>|@jml set owner to spiv|
|-|-|

spiv,

I couldn't reproduce the bug as defined. I even tried with this:
```
import unittest

class TestFoo(unittest.TestCase):
    def test_foo(self):
        pass


def test_suite():
    return unittest.TestSuite([TestFoo('test_foo'), TestFoo('test_foo')])
```

However, this was after I cleaned up a lot of cruft surrounding `TestVisitor` and dry run.

This cleanup is ready for review in dryrun-2024.

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)<a name="note_2"></a>|@jerub set owner to @jml|
|-|-|

There are conflicts when I attempt to merge this branch.
```
M      trial/test/test_test_visitor.py
C      trial/test/test_runner.py
M      trial/runner.py
C      trial/unittest.py
```

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_3"></a>|@jml set owner to @jerub|
|-|-|

back to you in dryrun-2024-2

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

twisted.trial.runner.PyUnitTestCase.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @jml<br>@exarkun set status to `new`|
|-|-|

twisted.trial.runner.PyUnitTestCase.visit's docstring, previously rather uninformative, is now incorrect.  Some further explanation might be merited here.

twisted.trial.unittest.TestVisitor has no class docstring.  One explaining the purpose of the class would be beneficial.

Less importantly:

  * Maintainer/Author comments would be more useful as @markup in a module docstring
  * pyflakes reports a couple things in ```twisted/trial/test/test_runner.py```, both of which may be new as a result of recent improvements to pyflakes, rather than introduced by this branch
  * the factoring of twisted.trial.test.test_runner.TestDryRun seems odd to me in a few ways:
    * test_empty doesn't even use the self.test created by setUp
    * test_pyunitTestCase covers one aspect of dry run support for pyunit TestCases, but it doesn't test that the tests aren't actually run, nor the behavior in the face of an empty suite.  Presumably the behaviors should be the same as for a trial TestCase, in which case I would define the previous three tests in a mixin and implement two subclasses, one of which sets up a trial TestCase and the other of which sets up a pyunit TestCase.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_6"></a>|@jml set owner to @exarkun|
|-|-|

I thought I assigned this for re-review. I guess I hit "preview" instead of "submit" again.

Fixed the test factoring and the pyflakes warnings (one was a false positive). Didn't do maintainer/author epytext markup. 

Ended up replacing TestVisitor with a callable.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @jml|
|-|-|

 * testNames is still duplicated in test_loader and test_scripts, but the comment to this effect has been deleted.
 * !__call!__ kind of sucks, I generally prefer a method with an informative name.  I'll leave this change to your discretion, though.
 * DryRunTest and PyUnitDryRunTest look good, but please give each a docstring.

All else looks good.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_8"></a>|@jml set status to `closed`|
|-|-|

(In [18983]) Make sure --dry-run treats pyunit tests well.
 * Author: jml
 * Reviewer: exarkun
 * Fixes [#2024](https://github.com/twisted/twisted/issues/2024)

This branch removes the TestVisitor class, allowing any callable to be
used as a test case visitor.  Also cleans up tests and interfaces.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
